storage/remote/azuread/token_provider.go

```diff
@@ -0,0 +1,145 @@
+	INGESTION_PUBLIC_AUDIENCE     = "https://monitor.azure.com//.default"
+)
+
+// tokenProvider is used to store and retrieve Azure AD accessToken.
+type tokenProvider struct {
+	// token is member used to store the current valid accessToken.
+	token string
+	// mu guards access to token
+	mu  sync.Mutex
+	ctx context.Context
+	// refreshTime is used to store the refresh time of the current valid accessToken.
+	refreshTime time.Time
+	// credentialClient is the Azure AD credential client that is being used to retrive accessToken.
+	credentialClient azcore.TokenCredential
+	options          *policy.TokenRequestOptions
+}
+
+// TokenProvider is the interface for Credential client.
+type TokenProvider interface {
+	getAccessToken() (string, error)
+}
+
+// newTokenProvider helps to fetch accessToken for different types of credential. This also takes care of
+// refreshing the accessToken before expiry. This accessToken is attached to the Authorization header while making requests.
+func newTokenProvider(ctx context.Context, cfg *AzureADConfig, cred azcore.TokenCredential) (TokenProvider, error) {
+	audience, err := getAudience(cfg.Cloud)
+	if err != nil {
+		return nil, err
+	}
+
+	tokenProvider := &tokenProvider{
+		ctx:              ctx,
+		credentialClient: cred,
+		options:          &policy.TokenRequestOptions{Scopes: []string{audience}},
+	}
+
+	_, err = tokenProvider.getAccessToken()
+	if err != nil {
+		return nil, errors.New("Failed to get access token: " + err.Error())
+	}
+
+	return tokenProvider, nil
+}
+
+// getAccessToken returns the current valid accessToken.
+func (tokenProvider *tokenProvider) getAccessToken() (string, error) {
+	tokenProvider.mu.Lock()
+	defer tokenProvider.mu.Unlock()
+	if tokenProvider.valid() {
+		return tokenProvider.token, nil
+	}
+	err := tokenProvider.getToken()
+	if err != nil {
+		return "", err
+	}
+	return tokenProvider.token, nil
+}
+
+// valid checks if the token in the token provider is valid and not expired
+func (tokenProvider *tokenProvider) valid() bool {
+	if tokenProvider.token != "" && tokenProvider.refreshTime.After(time.Now().UTC()) {
+		return true
+	} else {
+		return false
+	}
+}
+
+// getToken retrieves a new accessToken and stores the newly retrieved token in the tokenProvider.
+func (tokenProvider *tokenProvider) getToken() error {
+	accessToken, err := tokenProvider.credentialClient.GetToken(tokenProvider.ctx, *tokenProvider.options)
+	if err != nil {
+		return err
+	}
+
+	tokenProvider.token = accessToken.Token
+	err = tokenProvider.updateRefreshTime(accessToken)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// updateRefreshTime handles logic to set refreshTime. The refreshTime is set at half the duration of the actual token expiry.
+func (tokenProvider *tokenProvider) updateRefreshTime(accessToken azcore.AccessToken) error {
+	if len(accessToken.Token) == 0 {
+		return errors.New("Access Token is empty")
+	}
```

Comment:
I think it would be more logical to have this check take place after the token is fetched in `getToken()`, instead of both here and also in `valid()`

Reply:
done